core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java

```diff
@@ -0,0 +1,160 @@
+package alluxio.grpc;
+
+import alluxio.resource.LockResource;
+import com.google.common.base.Verify;
+import io.grpc.Channel;
+import io.grpc.ManagedChannel;
+import io.grpc.netty.NettyChannelBuilder;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Used to maintain singleton gRPC {@link ManagedChannel} instances per process.
+ *
+ * This class is used internally by {@link GrpcChannelBuilder} and {@link GrpcChannel}.
+ */
+public class GrpcManagedChannelPool {
+  /** Channels per address. */
+  @GuardedBy("mLock")
+  private static HashMap<ChannelKey, ManagedChannelReference> mChannels;
+  /** Used to control access to mChannel */
+  private static ReentrantReadWriteLock mLock;
+
+  static {
+    mChannels = new HashMap<>();
+    mLock = new ReentrantReadWriteLock(true);
+  }
```

Comment:
Right, I think that makes the code difficult to maintain. Could we provide a channel map per managed pool? How many managed pools do we expect to be making?

Reply:
Right now, it's a map per channel pool.
If we go with dynamic pool allocations then each call to GrpcChannelBuilder would need to provide that channel. I don't exactly know from how many points we create a Grpchannel so that's why I'm having a singleton pool instance.
@bf8086 WDYT?